packages/ref-core/src/ref_core/constraints.py

```diff
@@ -0,0 +1,148 @@
+from typing import Protocol, runtime_checkable
+
+import pandas as pd
+from attrs import frozen
+from loguru import logger
+
+from ref_core.exceptions import ConstraintNotSatisfied
+
+
+@runtime_checkable
+class GroupValidator(Protocol):
+    """
+    A constraint that must be satisfied when executing a given metric run.
+
+    All constraints must be satisfied for a given group to be run.
+    """
+
+    def validate(self, group: pd.DataFrame) -> bool:
+        """
+        Validate if the constraint is satisfied by the dataset.
+
+        This is executed after the apply method to determine if the constraint is satisfied.
+        If the constraint is not satisfied, the group will not be executed.
+
+        Parameters
+        ----------
+        group
+            A group of datasets that is being validated.
+
+        Returns
+        -------
+        :
+            Whether the constraint is satisfied
+        """
+        ...
+
+
+@runtime_checkable
+class GroupOperation(Protocol):
+    """
+    An operation to perform on a group of datasets resulting in a new group of datasets.
+    """
+
+    def apply(self, group: pd.DataFrame, data_catalog: pd.DataFrame) -> pd.DataFrame:
+        """
+        Perform an operation on the group of datasets.
+
+        If this operation modifies the group, the modified group should be returned.
+        If this operation fails, a ConstraintNotSatisfied exception should be raised.
+
+        Parameters
+        ----------
+        group
+            A group of datasets that is being validated.
+        data_catalog
+            The data catalog of datasets
+
+        Raises
+        ------
+        ConstraintNotSatisfied
+            The operation was not successful
+
+        Returns
+        -------
+        :
+            The updated group of datasets
+        """
+        ...
+
+
+Constraint = GroupOperation | GroupValidator
+"""
+A constraint that must be satisfied when executing a given metric run.
+
+This can include operations that are applied to a group of datasets which may modify the group,
+but may also include validators that check if the group satisfies a certain condition.
+
+All constraints must be satisfied for a given group to be run.
+If any constraint is not satisfied, the group will not be executed.
+"""
```

Comment:
Do I understand correctly?

A Constraint operates on a Group of datasets (which contains metadata about datasets only, no data) and returns a constrained Group of datasets, which is a subset of the input Group of datasets. Alternatively, it can signal that the input Group is not valid at all.

As a special case, a GroupValidator will not constrain the input Group of datasets, instead it can only signal that the input Group is not valid. Every GroupValidator could be implemented as a GroupOperation instead, but a GroupValidator might be easier to understand.

Reply:
I would write it as:

A Constraint operates on a Group of datasets (which contains metadata about datasets only, no data). There are two types of functionality that a Constraint may implement:
* A GroupOperation modifies a Group of datasets, potentially including additional datasets from the complete data catalog
* A GroupValidator can be used to check if the Group of datasets should be executed.

A Constraint can implement one or both of those interfaces in which case the GroupOperation is performed before the GroupValidator. Any Constraints that fail validation or raise exceptions will cause the Group of datasets to not be marked for execution.

> Every GroupValidator could be implemented as a GroupOperation instead, but a GroupValidator might be easier to understand.

Correct. If the GroupOperation raises a `ConstraintNotSatisfied` it would effectively be the same as a GroupValidator. Clearly delineating mutating the Group of datasets from validation feels nicer to me, but it might be adding unnecessary complexity.